tests/unit/test_security_middleware.py

```
@@ -0,0 +1,415 @@
+        assert "1.1.1.1" in middleware.requests
+
+    @pytest.mark.unit
+    def test_no_client_uses_unknown_ip(self) -> None:
+        """Verify requests without client info use 'unknown' as fallback IP."""
+        from fragrance_rater.middleware.security import RateLimitMiddleware
+
+        app = FastAPI()
+        app.add_middleware(RateLimitMiddleware, requests_per_minute=60, burst_size=10)
+
+        @app.get("/test")
+        async def _test_endpoint():
```

Ruff (0.15.12)

[warning] Missing return type annotation for private function `_test_endpoint`

(ANN202)